panic in noms log: Unrecognized type in diff function

```
$ noms log http://localhost:8000::fs1
commit ph4vjv428sula79b9ohuhcrvcgpnfni4
Parent: o38hugtf3l1e8rqtj89mijj1dq57eh4m
.root.contents.entries {
-   ".DS_Store": Inode {
-     attr: Attr {
-       ctime: 1.4699507797949877e+09,
-       gid: 502,
-       mode: 33188,
-       mtime: 1.4699507798150566e+09,
-       uid: 110853,
...

panic: Unrecognized type in diff function [recovered]
        panic: Unrecognized type in diff function

goroutine 32 [running]:
panic(0x4c12e0, 0xc82866c330)
        /opt/local/lib/go/src/runtime/panic.go:481 +0x3e6
github.com/attic-labs/noms/go/d.recoverWrappedTypes(0xc828215668, 0x0, 0x0, 0x0)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/go/d/try.go:146 +0xf2
panic(0x4c12e0, 0xc82866c330)
        /opt/local/lib/go/src/runtime/panic.go:443 +0x4e9
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8281eba40, 0x6, 0x6, 0xec14f0, 0xc82866c320, 0xec1540, 0xc8200ddf50, 0xec1540, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:38 +0x3ae
github.com/attic-labs/noms/cmd/noms/diff.diffStructs(0xec1a70, 0xc8282921a0, 0xc8281f2370, 0x5, 0x5, 0xc8200e16b0, 0x1, 0x1, 0xc8282064c0, 0xc8282d2c20, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:151 +0x973
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8281f2370, 0x5, 0x5, 0xec14f0, 0xc82866c2f0, 0xec17b8, 0xc8200ddf80, 0xec17b8, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:36 +0x665
github.com/attic-labs/noms/cmd/noms/diff.diffStructs(0xec1a70, 0xc8282921a0, 0xc8200119c0, 0x4, 0x4, 0xc8282d2ae0, 0x2, 0x2, 0xc828207640, 0xc8282d2c40, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:151 +0x973
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8200119c0, 0x4, 0x4, 0xec14f0, 0xc82000e490, 0xec17b8, 0xc8200ddfb0, 0xec17b8, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:36 +0x665
github.com/attic-labs/noms/cmd/noms/diff.diffMaps(0xec1a70, 0xc8282921a0, 0xc828290420, 0x3, 0x3, 0xec1908, 0xc82016a500, 0xc8200ee900, 0xec1908, 0xc820010580, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:112 +0x5ca
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc828290420, 0x3, 0x3, 0xec14f0, 0xc8281b8150, 0xec15f0, 0xc8200ee920, 0xec15f0, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:32 +0x354
github.com/attic-labs/noms/cmd/noms/diff.diffStructs(0xec1a70, 0xc8282921a0, 0xc8200ccaa0, 0x2, 0x2, 0xc8200e02c0, 0x1, 0x1, 0xc8281c8f40, 0xc8200ee940, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:151 +0x973
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8200ccaa0, 0x2, 0x2, 0xec14f0, 0xc8281b8120, 0xec17b8, 0xc8200dc840, 0xec17b8, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:36 +0x665
github.com/attic-labs/noms/cmd/noms/diff.diffStructs(0xec1a70, 0xc8282921a0, 0xc8281b80e0, 0x1, 0x1, 0xc8200ee760, 0x2, 0x2, 0xc828207640, 0xc8200ee960, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:151 +0x973
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8281b80e0, 0x1, 0x1, 0xec14f0, 0xc8281b80f0, 0xec17b8, 0xc8200dc870, 0xec17b8, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:36 +0x665
github.com/attic-labs/noms/cmd/noms/diff.diffStructs(0xec1a70, 0xc8282921a0, 0xc8282155a8, 0x0, 0x0, 0xc8200e0220, 0x1, 0x1, 0xc828207780, 0xc8200ee980, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:151 +0x973
github.com/attic-labs/noms/cmd/noms/diff.diff(0xec1a70, 0xc8282921a0, 0xc8282155a8, 0x0, 0x0, 0x0, 0x0, 0xec17b8, 0xc8200dc8a0, 0xec17b8, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:36 +0x665
github.com/attic-labs/noms/cmd/noms/diff.Diff.func1()
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:21 +0xe5
github.com/attic-labs/noms/go/d.Try(0xc828215678, 0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/go/d/try.go:55 +0x7e
github.com/attic-labs/noms/cmd/noms/diff.Diff(0xec1a70, 0xc8282921a0, 0xec17b8, 0xc8200dc8a0, 0xec17b8, 0xc82000be90, 0x0, 0x0)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/diff/diff.go:22 +0x94
main.writeDiffLines(0xe4421daf431fd1c0, 0x6e614e69139a5d6f, 0x9644778a, 0x5296, 0xc8281c8340, 0xc8200ee080, 0xc82000a480, 0x3, 0x3, 0xc828207dc0, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/noms_log.go:268 +0x628
main.printCommit(0xe4421daf431fd1c0, 0x6e614e69139a5d6f, 0x9644778a, 0x5296, 0xc8281c8340, 0xc8200ee080, 0xc82000a480, 0x3, 0x3, 0xc828207dc0, ...)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/noms_log.go:163 +0xefe
main.runLog.func1(0x6566c0, 0xc8281e2000, 0x0, 0x0)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/cmd/noms/noms_log.go:85 +0x155
github.com/attic-labs/noms/go/util/orderedparallel.New.func3(0xc8281eac60, 0xc8282ba240, 0xc828227c20, 0xc8281c5358, 0xc8281c5368, 0xc8281c5360, 0xc8281eacc0)
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/go/util/orderedparallel/ordered_parallel.go:51 +0x9f
created by github.com/attic-labs/noms/go/util/orderedparallel.New
        /Users/ahl/gosrc/src/github.com/attic-labs/noms/go/util/orderedparallel/ordered_parallel.go:61 +0x2e7
```

Dead reproducible. I can send in my dataset (once I get a new token (because I'm an idiot)).

--- a/cmd/noms/diff/diff.go
+++ b/cmd/noms/diff/diff.go
@@ -9,11 +9,12 @@ import (
 
 	"github.com/attic-labs/noms/go/d"
 	"github.com/attic-labs/noms/go/types"
+	"github.com/dustin/go-humanize"
 )
 
 func shouldDescend(v1, v2 types.Value) bool {
 	kind := v1.Type().Kind()
-	return !types.IsPrimitiveKind(kind) && kind == v2.Type().Kind()
+	return !types.IsPrimitiveKind(kind) && kind == v2.Type().Kind() && kind != types.RefKind
 }
 
 func Diff(w io.Writer, v1, v2 types.Value) error {
@@ -229,7 +230,13 @@ func write(w io.Writer, b []byte) {
 }
 
 func writeEncodedValue(w io.Writer, v types.Value) {
-	d.PanicIfError(types.WriteEncodedValue(w, v))
+	if v.Type().Kind() == types.BlobKind {
+		w.Write([]byte("Blob ("))
+		w.Write([]byte(humanize.Bytes(v.(types.Blob).Len())))
+		w.Write([]byte(")"))
+	} else {
+		d.PanicIfError(types.WriteEncodedValue(w, v))
+	}
 }
 
 func writeEncodedValueWithTags(w io.Writer, v types.Value) {

--- a/cmd/noms/diff/diff_test.go
+++ b/cmd/noms/diff/diff_test.go
@@ -5,9 +5,11 @@
 package diff
 
 import (
+	"strings"
 	"testing"
 
 	"github.com/attic-labs/noms/go/types"
+	"github.com/attic-labs/noms/go/util/test"
 	"github.com/attic-labs/testify/assert"
 	"github.com/syndtr/goleveldb/leveldb/util"
 )
@@ -203,4 +205,45 @@ func TestNomsListDiff(t *testing.T) {
 	Diff(buf, l1, l2)
 
 	assert.Equal(expected, buf.String())
+}
+
+func TestNomsBlobDiff(t *testing.T) {
+	assert := assert.New(t)
+
+	expected := "-   Blob (2.0 kB)\n+   Blob (11 B)\n"
+	b1 := types.NewBlob(strings.NewReader(strings.Repeat("x", 2*1024)))
+	b2 := types.NewBlob(strings.NewReader("Hello World"))
+	buf := util.NewBuffer(nil)
+	Diff(buf, b1, b2)
+	assert.Equal(expected, buf.String())
+}
+
+func TestNomsTypeDiff(t *testing.T) {
+	assert := assert.New(t)
+
+	expected := "-   List<Number>\n+   List<String>\n"
+	t1 := types.MakeListType(types.NumberType)
+	t2 := types.MakeListType(types.StringType)
+	buf := util.NewBuffer(nil)
+	Diff(buf, t1, t2)
+	assert.Equal(expected, buf.String())
+
+	expected = "-   List<Number>\n+   Set<String>\n"
+	t1 = types.MakeListType(types.NumberType)
+	t2 = types.MakeSetType(types.StringType)
+	buf = util.NewBuffer(nil)
+	Diff(buf, t1, t2)
+	assert.Equal(expected, buf.String())
+}
+
+func TestNomsRefDiff(t *testing.T) {
+	expected := "-   fckcbt7nk5jl4arco2dk7r9nj7abb6ci\n+   i7d3u5gekm48ot419t2cot6cnl7ltcah\n"
+	l1 := createList(1)
+	l2 := createList(2)
+	r1 := types.NewRef(l1)
+	r2 := types.NewRef(l2)
+	buf := util.NewBuffer(nil)
+	Diff(buf, r1, r2)
+
+	test.EqualsIgnoreHashes(t, expected, buf.String())
 }

